cmd/metal-api/internal/fsm/fsm.go

```diff
@@ -0,0 +1,262 @@
+		},
+		Dst: metal.ProvisioningEventWaiting.String(),
+	},
+	{
+		Name: metal.ProvisioningEventInstalling.String(),
+		Src: []string{
+			metal.ProvisioningEventWaiting.String(),
+		},
+		Dst: metal.ProvisioningEventInstalling.String(),
+	},
+	{
+		Name: metal.ProvisioningEventBootingNewKernel.String(),
+		Src: []string{
+			metal.ProvisioningEventInstalling.String(),
+		},
+		Dst: metal.ProvisioningEventBootingNewKernel.String(),
+	},
+	{
+		Name: metal.ProvisioningEventPhonedHome.String(),
+		Src: []string{
+			metal.ProvisioningEventBootingNewKernel.String(),
+			metal.ProvisioningEventPhonedHome.String(),
+			metal.ProvisioningEventMachineReclaim.String(),
+		},
+		Dst: metal.ProvisioningEventPhonedHome.String(),
+	},
+	{
+		Name: metal.ProvisioningEventPhonedHome.String(),
+		Src: []string{
+			metal.ProvisioningEventPlannedReboot.String(),
+		},
+		Dst: metal.ProvisioningEventPlannedReboot.String(),
+	},
+	{
+		Name: metal.ProvisioningEventPlannedReboot.String(),
+		Src: []string{
+			metal.ProvisioningEventPXEBooting.String(),
+			metal.ProvisioningEventPreparing.String(),
+			metal.ProvisioningEventRegistering.String(),
+			metal.ProvisioningEventWaiting.String(),
+			metal.ProvisioningEventInstalling.String(),
+			metal.ProvisioningEventBootingNewKernel.String(),
+			metal.ProvisioningEventPhonedHome.String(),
+		},
+		Dst: metal.ProvisioningEventPlannedReboot.String(),
+	},
+	{
+		Name: metal.ProvisioningEventMachineReclaim.String(),
+		Src: []string{
+			metal.ProvisioningEventPXEBooting.String(),
+			metal.ProvisioningEventPreparing.String(),
+			metal.ProvisioningEventRegistering.String(),
+			metal.ProvisioningEventWaiting.String(),
+			metal.ProvisioningEventInstalling.String(),
+			metal.ProvisioningEventBootingNewKernel.String(),
+			metal.ProvisioningEventPhonedHome.String(),
+		},
+		Dst: metal.ProvisioningEventMachineReclaim.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventPreparing.String(),
+		},
+		Dst: metal.ProvisioningEventPreparing.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventRegistering.String(),
+		},
+		Dst: metal.ProvisioningEventRegistering.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventWaiting.String(),
+		},
+		Dst: metal.ProvisioningEventWaiting.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventInstalling.String(),
+		},
+		Dst: metal.ProvisioningEventInstalling.String(),
+	},
+	{
+		Name: metal.ProvisioningEventAlive.String(),
+		Src: []string{
+			metal.ProvisioningEventBootingNewKernel.String(),
+		},
+		Dst: metal.ProvisioningEventBootingNewKernel.String(),
+	},
+}
+
+func ProvisioningEventForMachine(log *zap.SugaredLogger, ec *metal.ProvisioningEventContainer, event *metal.ProvisioningEvent) (*metal.ProvisioningEventContainer, error) {
```

Comment:
Can we add a comment to this func and generally explain the idea of calling it? Maybe even rename this function to `HandleProvisioningEvent`.

And for documentation something like: 

```golang
// HandleProvisioningEvent can be called to determine whether the given incoming event follows an expected lifecycle of a machine considering the event history of the given provisioning event container.
//
// The function returns a new provisioning event container that can then be persisted in the database. If an error is returned, the incoming event is not supposed to be persisted in the database.
//
// Among other things, this function can detect crash loops or other irregularities within a machine lifecycle and enriches the returned provisioning event container with this information. 
```